### PR TITLE
Proposal for hwb support in the sass:color module

### PR DESCRIPTION
**Pull request by mirisuzanne**
See issue #2834

### CODE REVIEW

File: proposal/color-4-hwb.md
Comment (nex3) [inline proposal/color-4-hwb.md]: Nit

```suggestion
The [CSS Color Module level 4][color-4] provides several new CSS formats for
```

File: proposal/color-4-hwb.md
Comment (nex3) [inline proposal/color-4-hwb.md]: Nit

```suggestion
This proposal defines a Sassified version of the `hwb()` color function
```

File: proposal/color-4-hwb.md
Comment (nex3) [inline proposal/color-4-hwb.md:141]: If this isn't a global function, why accept the global function color syntax?

File: proposal/color-4-hwb.md
Comment (nex3) [inline proposal/color-4-hwb.md]: I think this warrants more elaboration. Can you link to the algorithm for computing this?

File: proposal/color-4-hwb.md
Comment (nex3) [inline proposal/color-4-hwb.md:127]: One thing we haven't explicitly specified yet is what Sass's spec representation is of colors that were originally specified in different formats. In Dart Sass, we store colors as HSL values if they were originally HSL and RGB if they were originally RGB, lazily converting to the other format if that becomes necessary. Presumably we'll do something similar for HWB. This can be visible to the user—we should be explicit about whether, for example, `whiteness(hwb(..., $x, ...))` is guaranteed to return *exactly* `$x` or merely *approximately* `$x`.

I think the actual way you've specified this, treating sRGB values as the internal representation, is reasonable... but we should add an explicit amount of flexibility to enable the sort of optimization Dart Sass does.

You've also got to specify what the function actually returns :smile:.

File: proposal/color-4-hwb.md
Comment (nex3) [inline proposal/color-4-hwb.md]: Please describe the process of normalizing these in more detail, ideally with a formula.

File: proposal/color-4-hwb.md
Comment (nex3) [inline proposal/color-4-hwb.md:210]: These functions should also be more explicit about what they actually do.

To give some background: because writing a full language spec from scratch would take way too much time, we sort of fill it in piecemeal as it becomes relevant to each new proposal. This means that proposals like this need to do a bit of extra legwork to specify at least enough of the existing behavior to clearly specify the new behavior in relation to it.

**Review by nex3 — CHANGES_REQUESTED** [review]
Before we land this, I want to talk about our plan for `hwb()` going forward. Although it doesn't quite come out and say it explicitly, this proposal seems to imply that we won't ever add special Sass processing for the global `hwb()` function, and I want to push back on that.

Generally speaking, we're willing to add special support for new CSS syntax once it's shipped in at least one browser, since that's a pretty good guarantee that it's not going to completely change out from under us. So we probably *will* want to support global `hwb()` syntax in the future. This leads me to think that this proposal should explicitly address the following questions:

* Why add `color.hwb()` now rather than waiting for `hwb()` to land in a browser and implementing the global function when that happens?

* Once we do add support for global `hwb()`, what's the plan for `color.hwb()`? Do we deprecate it and perform a migration?

**Comment by nex3** [comment]
@mirisuzanne I made some revisions that I'd love your feedback on. In particular, I changed the design decisions section pretty substantially. The actual *decisions* are the same, but the reasoning behind them and the plans going forward are not.

File: proposal/color-4-hwb.md
Comment (mirisuzanne) [inline proposal/color-4-hwb.md:110]: CSS accepts several different `<angle>` units. I don't know how much they are used (`deg` is certainly the most common), but we might want to match? https://www.w3.org/TR/css3-values/#angle-value

File: proposal/color-4-hwb.md
Comment (mirisuzanne) [inline proposal/color-4-hwb.md:135]: Returning values from a function is totally overrated 😂

**Review by mirisuzanne — COMMENTED** [review]
@nex3 Changes and expansions here are great, thanks. Just one question below about hue angles…

File: proposal/color-4-hwb.md
Comment (nex3) [reply proposal/color-4-hwb.md:110]: We should support that, but it's tangential enough that I don't think it should be part of this proposal. I've filed https://github.com/sass/sass/issues/2904 to track that.

File: proposal/color-4-hwb.md
Comment (Awjin) [inline proposal/color-4-hwb.md:93]: What happens if `max < number`?

File: proposal/color-4-hwb.md
Comment (Awjin) [inline proposal/color-4-hwb.md:259]: Here and below (lines 265, 268, 347): should these have unit `%`?

**Review by Awjin — CHANGES_REQUESTED** [review]

File: proposal/color-4-hwb.md
Comment (nex3) [reply proposal/color-4-hwb.md:93]: In practice it never will be, since `number` always comes from a color channel which is guaranteed to be between 0 and `max`. I'll add a note to that effect.

File: proposal/color-4-hwb.md
Comment (nex3) [reply proposal/color-4-hwb.md:259]: Unfortunately, this represents the current behavior. We currently ignore units for saturation and lightness no matter what, which isn't ideal but isn't necessarily enough of a problem to warrant a deprecation pass either.

**Review by Awjin — APPROVED** [review]

File: proposal/color-4-hwb.md
Comment (jathak) [inline proposal/color-4-hwb.md]:
```suggestion
  * Let `blackness` be `$blackness / 100%`.
```

File: proposal/color-4-hwb.md
Comment (jathak) [inline proposal/color-4-hwb.md]:
```suggestion
    `whiteness`, and `blackness` [to RGB][].
```

File: proposal/color-4-hwb.md
Comment (jathak) [inline proposal/color-4-hwb.md]:
```suggestion
  * Set `alpha` to `alpha + $alpha` clamped between 0 and 1.
```

File: proposal/color-4-hwb.md
Comment (jathak) [inline proposal/color-4-hwb.md]:
```suggestion
  * If either `$whiteness` or `$blackness` aren't either null or numbers with
```

**Review by jathak — APPROVED** [review]

**Review by nex3 — APPROVED** [review]

**Pull request merged**
